ghcide/src/Development/IDE/Core/WorkerThread.hs

```diff
@@ -0,0 +1,46 @@
+module Development.IDE.Core.WorkerThread
+    (withWorkerQueue, blockRunInThread)
+ where
+
+import           Control.Concurrent.Async
+import           Control.Concurrent.STM
+import           Control.Concurrent.Strict (newBarrier, signalBarrier,
+                                            waitBarrier)
+import           Control.Monad             (forever)
+import           Control.Monad.Cont        (ContT (ContT))
+
+-- Note [Serializing runs in separate thread]
+-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+-- We often want to take long-running actions using some resource that cannot be shared.
+-- In this instance it is useful to have a queue of jobs to run using the resource.
+-- Like the db writes, session loading in session loader, shake session restarts.
+--
+-- Originally we used various ways to implement this, but it was hard to maintain and error prone.
+-- Moreover, we can not stop these threads uniformly when we are shutting down the server.
+--
+-- `Development.IDE.Core.WorkerThread` module provides a simple api to implement this easily.
+-- * `withWorkerQueue`: accepts an action to run in separate thread and returns a `TQueue` to send the actions to run.
+-- * `blockRunInThread` : accepts a `TQueue` and an action to run in separate thread and waits for the result.
+
+
+-- | withWorkerQueue creates a new TQueue and runs the workerAction in a separate thread.
+withWorkerQueue :: (t -> IO a) -> ContT () IO (TQueue t)
+withWorkerQueue workerAction = ContT $ \mainAction -> do
+    q <- newTQueueIO
+    withAsync (writerThread q) $ \_ -> mainAction q
+    where
+        writerThread q =
+            forever $ do
+                l <- atomically $ readTQueue q
+                workerAction l
+
+-- | blockRunInThread run and wait for the result
+blockRunInThread :: TQueue (IO ()) -> IO result -> IO result
```

Comment:
Subjective, but I think we more often use "await" rather than "block" in this sort of context. So maybe call it `awaitRunInThread`?

Reply:
Yep